server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchFieldsPhase.java

```diff
@@ -61,6 +63,19 @@ public FetchSubPhaseProcessor getProcessor(FetchContext fetchContext) {
             : fetchFieldsContext.fields().isEmpty() ? null
             : FieldFetcher.create(searchExecutionContext, fetchFieldsContext.fields());
 
+        final Set<FieldAndFormat> fetchContextMetadataFields = new HashSet<>();
+        if (fetchFieldsContext != null && fetchFieldsContext.fields() != null && fetchFieldsContext.fields().isEmpty() == false) {
+            for (final FieldAndFormat fieldAndFormat : fetchFieldsContext.fields()) {
+                if (SourceFieldMapper.NAME.equals(fieldAndFormat.field) || IdFieldMapper.NAME.equals(fieldAndFormat.field)) {
```

Comment:
I don't understand why source and id are involved here. Is this a different existing problem that they are returned in cases when they shouldn't be?

Reply:
@salvatore-campagna maybe add a comment explaining why this is necessary?

Reply:
`_source` and `_id` are treated differently because they are retrieved by other means and not by this logic...so we skip them to avoid duplicates. You can see similar logic in `StoredFieldsPhase`.

Reply:
If you look into `FieldVisitor` and how it is used you will see logic like the following

```
public void reset() {
        if (fieldsValues != null) {
            fieldsValues.clear();
        }
        source = null;
        id = null;

        requiredFields.addAll(BASE_REQUIRED_FIELDS);
        if (loadSource) {
            requiredFields.add(sourceFieldName);
        }
    }
```

```
BASE_REQUIRED_FIELDS
```

includes `_source` and `_id`. This is just to say that the responsibility of loading those fields is somewhere else (`StoredFieldLoader`).

Reply:
But how did it work when this code was not here?

Reply:
Note that some of the yaml tests I wrote actually work in main.

The ones failing is main are the following ones
```
* fetch stored fields wildcard => in main we return `_ignored_source` when `stored_fields: [*]`
* wildcard fields and stored fields combination => same as before
* fields with ignored source in fields => this one is failing because fields requested via `fields` and so in `FetchFieldContext` are fetched via `ValueFetcher` which uses a stored field but the `FiedlFetcher` for **non-metadata** fields put those values in `fields` and not in `metadata fields`. This is an issue because `_ignored_source` is a metadata field which ends up into doc fields
* fetch fields by name => fails for a reason similar to above
```

For the third test failing above see this

```
public void process(HitContext hitContext) throws IOException {
                final Map<String, DocumentField> fields = fieldFetcher != null
                    ? fieldFetcher.fetch(hitContext.source(), hitContext.docId())
                    : Collections.emptyMap();
                final Map<String, DocumentField> metadataFields = metadataFieldFetcher.fetch(hitContext.source(), hitContext.docId());
                hitContext.hit().addDocumentFields(fields, metadataFields);
            }
```

Note that there is a `fieldFetcher` for document fields and a `metadataFieldFetcher` for metadata fields.

@lkts does it make sense?

Reply:
Thank you, i understand now!